src/lib/updating-element.ts

```diff
@@ -0,0 +1,552 @@
+
+  /**
+   * Describes if setting a property should trigger invalidation and updating.
+   * This function takes the `newValue` and `oldValue` and returns `true` if
+   * invalidation should occur. If not present, a strict identity check
+   * (eg. === operator) is used. This is useful if a property should be
+   * considered dirty only if some condition is met, like if a key of an
+   * object value changes.
+   */
+  shouldInvalidate?(value: T, oldValue: T): boolean;
+
+}
+
+/**
+ * Object that describes accessors to be created on the element prototype.
+ * An accessor is created for each key with the given options.
+ */
+export interface PropertyDeclarations {
+  [key: string]: PropertyDeclaration;
+}
+
+type PropertyDeclarationMap = Map<PropertyKey, PropertyDeclaration>;
+
+type AttributeMap = Map<string, PropertyKey>;
+
+export type PropertyValues = Map<PropertyKey, unknown>;
+
+/**
+ * Decorator which creates a property. Optionally a `PropertyDeclaration` object
+ * can be supplied to describe how the property should be configured.
+ */
+export const property = (options?: PropertyDeclaration) => (proto: Object, name: string) => {
+  (proto.constructor as typeof UpdatingElement).createProperty(name, options);
+};
+
+// serializer/deserializers for boolean attribute
+const fromBooleanAttribute = (value: string) => value !== null;
+const toBooleanAttribute = (value: string) => value ? '' : null;
+
+export interface ShouldInvalidate {
+  (value: unknown, old: unknown): boolean;
+}
+
+/**
+ * Change function that returns true if `value` is different from `oldValue`.
+ * This method is used as the default for a property's `shouldInvalidate` function.
+ */
+export const notEqual: ShouldInvalidate = (value: unknown, old: unknown): boolean => {
+  // This ensures (old==NaN, value==NaN) always returns false
+  return old !== value && (old === old || value === value);
+};
+
+const defaultPropertyDeclaration: PropertyDeclaration = {
+  attribute: true,
+  type: String,
+  reflect: false,
+  shouldInvalidate: notEqual
+};
+
+const microtaskPromise = new Promise((resolve) => resolve(true));
+
+const STATE_HAS_UPDATED = 1;
+const STATE_IS_UPDATING = 1 << 2;
+const STATE_IS_REFLECTING = 1 << 3;
+type ValidationState = typeof STATE_HAS_UPDATED | typeof STATE_IS_UPDATING | typeof STATE_IS_REFLECTING;
+
+/**
+ * Base element class which manages element properties and attributes. When
+ * properties change, the `update` method is asynchronously called. This method
+ * should be supplied by subclassers to render updates as desired.
+ */
+export abstract class UpdatingElement extends HTMLElement {
+
+  /**
+   * Maps attribute names to properties; for example `foobar` attribute
+   * to `fooBar` property.
+   */
+  private static _attributeToPropertyMap: AttributeMap = new Map();
+
+  /**
+   * Marks class as having finished creating properties.
+   */
+  private static _finalized = true;
+
+  /**
+   * Memoized result of computed observedAttributes.
```

Comment:
Why does this have to be memoized? For `customElements.define` it's torn off at `define` time and remembered by the browser.

Reply:
Idea was for calls to `super.observedAttributes` in overrides but probably not needed. Can remove.